File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/CAShapeLayer动画/WGLineChartView.m
//
//  WGLineChartView.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/27.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGLineChartView.h"

@interface WGLineChartView()

@property (nonatomic, strong) CAShapeLayer *shapeLayer;

@end

@implementation WGLineChartView

@synthesize lineColor = _lineColor;

-(instancetype)initWithFrame:(CGRect)frame{
    self= [super initWithFrame:frame];
    if (self) {
        
        [self createView];
    }
    return self;
}

- (void)createView{
    
    self.shapeLayer = [[CAShapeLayer alloc] init];
    _shapeLayer.fillColor = UIColor.whiteColor.CGColor;
    _shapeLayer.strokeEnd = 0;
    _shapeLayer.lineWidth = 10;
    
    self.backgroundColor = UIColor.whiteColor;
    [self.layer addSublayer:_shapeLayer];
}

-(void)setPointArr:(NSArray *)pointArr{
    
    _pointArr = pointArr;
    if (pointArr.count == 0) {
        return;
    }
    UIBezierPath *path = [UIBezierPath bezierPath];
    for (NSValue *pointValue in pointArr) {
        
        CGPoint point = [pointValue CGPointValue];
        if ([pointValue isEqualToValue:pointArr[0]]) {
            
            [path moveToPoint:point];
        }else{
            [path addLineToPoint:point];
        }
    }
    
    _shapeLayer.path = path.CGPath;
    _shapeLayer.strokeColor = self.lineColor.CGColor;
}

- (void)startAni{
    
    CABasicAnimation *ani = [[CABasicAnimation alloc] init];
    ani.keyPath = @"strokeEnd";
    ani.fromValue = @(0);
    ani.toValue = @(1);
    ani.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseIn];
    ani.duration = 3;
    ani.autoreverses = NO;
    [ani setRemovedOnCompletion:NO];
    
    _shapeLayer.strokeEnd = 1;
    [_shapeLayer addAnimation:ani forKey:nil];
}

-(UIColor *)lineColor{
    if (!_lineColor) {
        _lineColor = [UIColor redColor];
    }
    return _lineColor;
}

-(void)setLineColor:(UIColor *)lineColor{
    _lineColor = lineColor;
    _shapeLayer.strokeColor = _lineColor.CGColor;
}

@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/CAShapeLayer动画/WGLineChartView.swift
//
//  WGLineChartView.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/5/7.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

class WGLineChartView: UIView {

    let lineChart = CAShapeLayer()
    let ani = CABasicAnimation()
    
    override init(frame: CGRect) {
        
        super.init(frame: frame)
        
        self.backgroundColor = UIColor.white
        lineChart.lineCap = .round
        lineChart.lineJoin = .round
        lineChart.fillColor = UIColor.white.cgColor
        lineChart.lineWidth = 10
        lineChart.strokeEnd = 0.0
        self.layer.addSublayer(lineChart)
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        
        fatalError("init(coder:) has not been implemented")
    }
    
    override func draw(_ rect: CGRect) {
        
        let bezierPath = UIBezierPath()
        bezierPath.lineCapStyle = .round
        bezierPath.lineJoinStyle = .round
        bezierPath.lineWidth = 10.0
        bezierPath.move(to: CGPoint(x: 30+35, y: 260))
        bezierPath.addLine(to: CGPoint(x: 100+35, y: 100))
        bezierPath.addLine(to: CGPoint(x: 170+35, y: 240))
        bezierPath.addLine(to: CGPoint(x: 240+35, y: 170))
        bezierPath.addLine(to: CGPoint(x: 310+35, y: 220))
        
        lineChart.path = bezierPath.cgPath
        lineChart.strokeColor = UIColor.red.cgColor
        
        ani.keyPath = "strokeEnd"
        ani.timingFunction = CAMediaTimingFunction(name: .easeIn)
        ani.fromValue = 0.0
        ani.toValue = 1.0
        ani.autoreverses = false
        ani.duration = 2.0
    }

    func chartAniStart() {
        
        lineChart.strokeEnd = 1.0
        lineChart.add(ani, forKey: nil)
    }
    
}
